### PR TITLE
JS: add indirect command injection sink for a concatenated array

### PR DESCRIPTION
**Pull request by erik-krogh**
This PR simply adds the sink for CVE-2018-6342. 

Previously we would flag: 
```
child_process.execFile("/bin/bash", ["-c", source]); // NOT OK
```

With this PR we also flag: 
```
child_process.execFile("/bin/bash", ["-c"].concat(source)); // NOT OK
```

### CODE REVIEW

**Comment by esbena** [comment]
This does not allow us to flag the case where we need to propagte through the receiver of the `concat` call, does it?

```js
child_process.execFile("/bin/bash", [source].concat("--foo")); // NOT OK
```

**Comment by esbena** [comment]
Perhaps we should look into exposing something like `TaintTracking::ArrayFunctionTaintStep` for this kind of additional type tracking - that would scale better than these ad hoc method models.

**Comment by erik-krogh** [comment]
> Perhaps we should look into exposing something like `TaintTracking::ArrayFunctionTaintStep` for this kind of additional type tracking - that would scale better than these ad hoc method models.

I've tried to do that for now (I additionally added "arr.concat(..)" as a taint step). 

I basically added the array-step to the backtracker in `argumentList` in `IndirectCommandArgument.qll`, and use that instead of the ad-hoc approach I had before.

File: javascript/ql/src/semmle/javascript/security/dataflow/IndirectCommandArgument.qll [hunk 1]
```diff
@@ -61,15 +64,22 @@ private DataFlow::SourceNode argumentList(SystemCommandExecution sys, DataFlow::
...
-  exists(
-    DataFlow::ArrayCreationNode args, DataFlow::Node shell, string dashC
-  |
+  exists(DataFlow::ArrayCreationNode args, DataFlow::Node shell, string dashC |
     shellCmd(shell.asExpr(), dashC) and
     shell = commandArgument(sys, DataFlow::TypeBackTracker::end()) and
-    args = argumentList(sys, DataFlow::TypeBackTracker::end()) and
     args.getAPropertyWrite().getRhs().mayHaveStringValue(dashC) and
-    source = args.getAPropertyWrite().getRhs()
+    args = argumentList(sys, DataFlow::TypeBackTracker::end()) and
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/IndirectCommandArgument.qll]: Nit: lets create and use the unary `argumentList(sys)` to reduce this `::end` duplication.

**Comment by erik-krogh** [comment]
The [evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-asger.northeurope.cloudapp.azure.com_1581451991662)  ([part2](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-asger.northeurope.cloudapp.azure.com_1581605701257)) looks surprisingly good.  
No new results, but performance is now better.

**Pull request merged**